src/main/java/com/redhat/devtools/lsp4ij/features/formatting/LSPClientSideOnTypeFormattingTypedHandler.java

```diff
@@ -50,120 +52,119 @@ public Result charTyped(char charTyped,
+        FormattingScope formatOnStatementTerminatorScope = formatOnStatementTerminatorScopeRef.getValue();
+        String formatOnStatementTerminatorCharacters = formatOnStatementTerminatorCharactersRef.getValue();
+        boolean formatOnCompletionTrigger = formatOnCompletionTriggerRef.getValue();
+        String formatOnCompletionTriggerCharacters = formatOnCompletionTriggerCharactersRef.getValue();
+
+        // Close braces
+        if (formatOnCloseBrace &&
                 // Make sure the formatter supports formatting of the configured scope
-                ((formattingFeature.getFormatOnStatementTerminatorScope(file) == FormattingScope.FILE) || rangeFormattingSupported)) {
-                String formatOnStatementTerminatorCharacters = formattingFeature.getFormatOnStatementTerminatorCharacters(file);
-                if (StringUtil.isNotEmpty(formatOnStatementTerminatorCharacters) &&
-                    (formatOnStatementTerminatorCharacters.indexOf(charTyped) > -1)) {
-                    return handleStatementTerminatorTyped(
+                (formatOnCloseBraceScope != null) &&
+                ((formatOnCloseBraceScope == FormattingScope.FILE) || rangeFormattingSupported)) {
+            Map.Entry<Character, Character> bracePair = ContainerUtil.find(
+                    LSPIJEditorUtils.getBracePairs(file).entrySet(),
+                    entry -> entry.getValue() == charTyped
+            );
+            if (bracePair != null) {
+                Character openBraceChar = bracePair.getKey();
+                Character closeBraceChar = bracePair.getValue();
+                if (StringUtil.isEmpty(formatOnCloseBraceCharacters) ||
+                        (formatOnCloseBraceCharacters.indexOf(closeBraceChar) > -1)) {
+                    return handleCloseBraceTyped(
                             project,
                             editor,
                             file,
-                            formattingFeature,
-                            charTyped
+                            formatOnCloseBraceScope,
+                            openBraceChar,
+                            closeBraceChar
                     );
                 }
             }
+        }
+
+        // Statement terminators
+        if (formatOnStatementTerminator &&
+                // Make sure the formatter supports formatting of the configured scope
+                (formatOnStatementTerminatorScope != null) &&
+                ((formatOnStatementTerminatorScope == FormattingScope.FILE) || rangeFormattingSupported)) {
+            if (StringUtil.isNotEmpty(formatOnStatementTerminatorCharacters) &&
+                    (formatOnStatementTerminatorCharacters.indexOf(charTyped) > -1)) {
+                return handleStatementTerminatorTyped(
+                        project,
+                        editor,
+                        file,
+                        formatOnStatementTerminatorScope,
```

Comment:
We were only passing `formattingFeature` to get to `formatOnStatementTerminatorScope`, so just pass that scope.